From: Alberto Garcia <berto@igalia.com>
Description: convert inline XPM icons to GdkPixbuf
 gdk-pixbuf starting from 2.42.11 disables the XPM loader by default.
 This is a temporary patch that replaces the XPM images in
 ui/gtk/pixmaps.c with their equivalents in the GdkPixbuf format, as
 generated by the gdk-pixbuf-csource tool.
Bug: https://sourceforge.net/p/fuse-emulator/bugs/509/
diff --git a/ui/gtk/browse.c b/ui/gtk/browse.c
index 2663b485..94d8330e 100644
--- a/ui/gtk/browse.c
+++ b/ui/gtk/browse.c
@@ -155,7 +155,7 @@ create_dialog( void )
   gtk_box_pack_start( GTK_BOX( content_area ), scrolled_window, TRUE, TRUE, 0 );
 
   /* The tape marker pixbuf */
-  tape_marker_pixbuf = gdk_pixbuf_new_from_xpm_data( gtkpixmap_tape_marker );
+  tape_marker_pixbuf = gdk_pixbuf_new_from_inline( -1, gtkpixmap_tape_marker, FALSE, NULL );
   /* FIXME: unref this at exit */
 
   /* And the list itself */
diff --git a/ui/gtk/statusbar.c b/ui/gtk/statusbar.c
index 36be3c7f..c4d592e1 100644
--- a/ui/gtk/statusbar.c
+++ b/ui/gtk/statusbar.c
@@ -59,29 +59,29 @@ gtkstatusbar_create( GtkBox *parent )
 
   /* FIXME: unref these pixbuf on statusbar destroy */
   pixbuf_tape_inactive =
-    gdk_pixbuf_new_from_xpm_data( gtkpixmap_tape_inactive );
+    gdk_pixbuf_new_from_inline( -1, gtkpixmap_tape_inactive, FALSE, NULL );
   pixbuf_tape_active =
-    gdk_pixbuf_new_from_xpm_data( gtkpixmap_tape_active );
+    gdk_pixbuf_new_from_inline( -1, gtkpixmap_tape_active, FALSE, NULL );
 
   pixbuf_mdr_inactive =
-    gdk_pixbuf_new_from_xpm_data( gtkpixmap_mdr_inactive );
+    gdk_pixbuf_new_from_inline( -1, gtkpixmap_mdr_inactive, FALSE, NULL );
   pixbuf_mdr_active =
-    gdk_pixbuf_new_from_xpm_data( gtkpixmap_mdr_active );
+    gdk_pixbuf_new_from_inline( -1, gtkpixmap_mdr_active, FALSE, NULL );
 
   pixbuf_disk_inactive =
-    gdk_pixbuf_new_from_xpm_data( gtkpixmap_disk_inactive );
+    gdk_pixbuf_new_from_inline( -1, gtkpixmap_disk_inactive, FALSE, NULL );
   pixbuf_disk_active =
-    gdk_pixbuf_new_from_xpm_data( gtkpixmap_disk_active );
+    gdk_pixbuf_new_from_inline( -1, gtkpixmap_disk_active, FALSE, NULL );
 
   pixbuf_pause_inactive =
-    gdk_pixbuf_new_from_xpm_data( gtkpixmap_pause_inactive );
+    gdk_pixbuf_new_from_inline( -1, gtkpixmap_pause_inactive, FALSE, NULL );
   pixbuf_pause_active =
-    gdk_pixbuf_new_from_xpm_data( gtkpixmap_pause_active );
+    gdk_pixbuf_new_from_inline( -1, gtkpixmap_pause_active, FALSE, NULL );
 
   pixbuf_mouse_inactive =
-    gdk_pixbuf_new_from_xpm_data( gtkpixmap_mouse_inactive );
+    gdk_pixbuf_new_from_inline( -1, gtkpixmap_mouse_inactive, FALSE, NULL );
   pixbuf_mouse_active =
-    gdk_pixbuf_new_from_xpm_data( gtkpixmap_mouse_active );
+    gdk_pixbuf_new_from_inline( -1, gtkpixmap_mouse_active, FALSE, NULL );
 
   speed_status = gtk_label_new( "100%" );
   gtk_label_set_width_chars( GTK_LABEL( speed_status ), 8 );
diff --git a/convert-xpm.sh b/convert-xpm.sh
new file mode 100755
index 00000000..52668c9b
--- /dev/null
+++ b/convert-xpm.sh
@@ -0,0 +1,18 @@
+#!/bin/sh -e
+
+cd ui/gtk
+
+echo '#include "gtkinternals.h"' > pixmaps-tmp.c
+
+for f in $(sed -n '/gtkpixmap_/s/.*gtkpixmap_\(.*\)\[\].*/\1/p' pixmaps.c); do
+    echo "Extracting $f.xpm"
+    echo '/* XPM */' > "$f.xpm"
+    sed -n "/gtkpixmap_$f/",'/^$/p' pixmaps.c >> "$f.xpm"
+    echo "Converting $f.xpm to $f.png"
+    ffmpeg -i "$f.xpm" "$f.png" 2> /dev/null
+    echo "Converting $f.png to GdkPixbuf"
+    gdk-pixbuf-csource --extern --raw --name="gtkpixmap_$f" "$f.png" >> pixmaps-tmp.c
+    sed -i "/gtkpixmap_$f\[/s/char \*/guint8 /" gtkinternals.h
+done
+
+mv pixmaps-tmp.c pixmaps.c
